kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,760 @@
+    val propertyList: List<Property>,
+    val originalElement: Element,
+    val name: String = fqClassName.substringAfter(packageName)
+        .replace('.', '_')
+        .removePrefix("_"),
+    val hasCompanionObject: Boolean,
+    val visibility: Visibility,
+    val elementUtils: Elements,
+    val genericTypeNames: List<TypeVariableName>?) {
+
+  fun generate(adapterName: String, fileSpecBuilder: FileSpec.Builder) {
+    val nameAllocator = NameAllocator()
+    fun String.allocate() = nameAllocator.newName(this)
+
+    val originalTypeName = originalElement.asType().asTypeName()
+    val moshiName = "moshi".allocate()
+    val moshiParam = ParameterSpec.builder(moshiName, Moshi::class).build()
+    val typesParam = ParameterSpec.builder("types".allocate(),
+        ParameterizedTypeName.get(ARRAY, Type::class.asTypeName())).build()
+    val reader = ParameterSpec.builder("reader".allocate(),
+        JsonReader::class).build()
+    val writer = ParameterSpec.builder("writer".allocate(),
+        JsonWriter::class).build()
+    val value = ParameterSpec.builder("value".allocate(),
+        originalTypeName.asNullable()).build()
+    val jsonAdapterTypeName = ParameterizedTypeName.get(JsonAdapter::class.asClassName(),
+        originalTypeName)
+
+    // Create fields
+    val adapterProperties = propertyList
+        .distinctBy { it.typeName to it.jsonQualifiers }
+        .associate { prop ->
+          val typeName = prop.typeName
+          val qualifierNames = prop.jsonQualifiers.joinToString("") {
+            "at${it.annotationType.asElement().simpleName.toString().capitalize()}"
+          }
+          val propertyName = typeName.simplifiedName().allocate().let {
+            if (qualifierNames.isBlank()) {
+              it
+            } else {
+              "$it$qualifierNames"
+            }
+          }.let { "${it}Adapter" }
+          val adapterTypeName = ParameterizedTypeName.get(JsonAdapter::class.asTypeName(), typeName)
+          val key = typeName to prop.jsonQualifiers
+          return@associate key to PropertySpec.builder(propertyName, adapterTypeName, PRIVATE)
+              .apply {
+                val qualifiers = prop.jsonQualifiers.toList()
+                val standardArgs = arrayOf(moshiParam,
+                    if (typeName is ClassName && qualifiers.isEmpty()) {
+                      ""
+                    } else {
+                      CodeBlock.of("<%T>",
+                          typeName)
+                    },
+                    typeName.makeType(elementUtils, typesParam, genericTypeNames ?: emptyList()))
+                val standardArgsSize = standardArgs.size + 1
+                val (initializerString, args) = when {
+                  qualifiers.isEmpty() -> "" to emptyArray()
+                  qualifiers.size == 1 -> {
+                    ", %${standardArgsSize}T::class.java" to arrayOf(
+                        qualifiers.first().annotationType.asTypeName())
```

Comment:
this looks like it will run into https://github.com/ansman/kotshi/issues/60 (same below)

Reply:
Interesting. Definitely interested in seeing Moshi support that case

Reply:
Moshi does support it. Both the KotlinJsonAdapter and the ClassJsonAdapter hand off the annotation instances to `adapter(Type, Set...)`.

Reply:
that's via reflectively pulling off the annotation instances, whereas the idea here was to avoid that since we know them upfront

Reply:
This PR does `adapter(Type, Class<?>`). This will create a new annotation instance that loses the declared annotation elements.

Reply:
I see what you mean, but I'm confused on what you mean by `Moshi does support it` above? One way I could envision that case working here is by reflectively pulling them off the parameters at runtime, as kotlin annotations don't allow for something like `AutoAnnotation` since they're `final` :/ . 

The other, albeit possibly crazy way, is to generate proxy invocation handlers that include the statically defined data (kind of our own inline kotlin-ified autoannotation with proxies)